Staging

cnuke opened the pull request.

Commit:
cbe: clean-up work in preparation for proving

This cleans up some internal aspects of the CBE with the goal of simplifying
later proving with gnatprove. The CBE library interface is not affected by
these changes.

The changes in detail
---------------------

* Remove outcommented code
* Get rid of null statements
* Get rid of invalid snapshot ID
* Get rid of Snapshot_ID_Invalid_Slot
* Get rid of Address_Type and "*_SPARK" sub-programs (better said, their
  non-"SPARK" counterparts)
* Static constant Block_Size
* Static constant Superblock_Nr_Of_Snapshots
* Clean up snapshot ID and index
* Clean up C++ Superblock_index
* Key_ID_Type/Key_ID_Storage_Type without mod

Commit:
cbe: decouple IO buffer management from CBE library

The IO buffer is allocated outside of and separately from the CBE library.
Whenever a call to the library requires access to the buffer it is handed in as
"in out" parameter. Wherever the CBE library interface allows for it, copying
data from or to the IO buffer inside the library is avoided. Instead, indices
into the buffer are communicated between the CBE library and its user. This
spares block copy-operations, simplifies interfaces, and allows the CBE
modules to concentrate on mere state keeping.

Changes to the CBE library interface
------------------------------------

* ``Execute``: Now receives the IO buffer as additional "in out" parameter.

* ``Obtain_Client_Data_2``: Now receives the IO buffer as additional "in out"
  parameter. (Note: this is a temporary API function that will be removed soon)

* ``Has_IO_Request``: Replaces former Has_IO_Data_To_Write and
  IO_Data_Required. In addition to the former procedures it also returns a data
  index. If the request is a "write", then the corresponding data is located in
  the IO-buffer slot referenced by the data index. Works similar to
  Crypto_Cipher/Plain_Data_Required for the crypto back-end.

* ``IO_Request_In_Progress``: Replaces former IO_Data_Read_In_Progress and
  Obtain_IO_Data (only the acknowledgement part). The new procedure takes only
  a data index as in parameter and acknowledges to the CBE that the outer world
  accepted the IO request behind the indicated IO-buffer slot and that the
  request is now in progress. Works similar to
  Crypto_Cipher/Plain_Data_Requested for the crypto back-end.

* ``IO_Request_Completed``: Replaces former Supply_IO_Data and
  Ack_IO_Data_To_Write. Takes a data index as in parameter for referencing the
  finished IO request and its IO-buffer slot, as well as a boolean whether the
  request was successful. If the request was a "read" and successful, the read
  data is expected to be present in the referenced IO-buffer slot. Works
  similar to Supply_Crypto_Cipher/Plain_Data for the crypto back-end.

cnuke committed:
cbe: snapshot interface

Adds sub-programs to the CBE library interface that can be used to manage
snapshots explicitely and adapts the client interface of the CBE library in a
way that one can read from any specific snapshot (the current snapshot can
still be written additionally). The interfaces for sealing generations and
securing superblocks were removed from the CBE library as the new snapshot
interface renders them useless.

New interface to manage snapshots
---------------------------------

:Library.Create_Snapshot:

Start creating a normal or a quarantine snapshot. Returns ID of new snapshot.

:Library.Snapshot_Creation_Complete:

Returns whether the snapshot behind a given ID was the last that was created
and whether the creation finished successful.

:Library.Active_Snapshot_IDs:

Returns an array of IDs of all existing snapshots that can be accessed
currently.

Changes to the existing client interface
----------------------------------------

:Library.Submit_Client_Request:

Received an additional "in" parameter that denotes the ID of the snapshot the
request refers to. A value of 0 has the special meaning that the request
refers to the current snapshot.

:Library.Is_Securing_Superblock:
:Library.Is_Sealing_Generation:
:Library.Start_Securing_Superblock:
:Library.Start_Sealing_Generation:

Were removed.

cnuke committed:
server/cbe_block: do not flag initial snapshot

We do not need to keep the initial snapshot. As it is of generation 0
all nodes will eventually changed in place.

Commit:
cbe: gnatprove in mode "check_all" succeeds

This required especially the adaption of sub-program interfaces according to
the requirements of SPARK (including some sub-programs of the CBE library
interface - see below) and the separation of debugging code into a dedicated
package with "SPARK_Mode => Off". To be able to have calls to the debugging
package in SPARK code without breaking proofs, an  ADC file was added
containing a global static switch that, by default, turns the evaluation of
code marked with "pragma Debug" off.

Changes to the CBE library interface
------------------------------------

:Supply_Client_Data:

Was a function, is now a procedure that provides its former return value
through a new "out" parameter.

:Drop_Completed_Request:

Removed request "in" parameter as the current implementation doesn't use it.

Commit:
.ci/cbe.gpr: show full paths on gnatprove messages

cnuke committed:
lib/cbe: use current query VBA to check for EOT

Using 'Curr_Query_Prim := Primitive.Invalid_Object' will also reset the
Block_Number that is needed in the check (stems from C++ code where
invalidating was done via setting the Operation to INVALID but keeping
reset of the object intact).

Commit:
cbe: get rid of Library.Obtain_Client_Data_2

The Obtain_Client_Data_2 sub-program (part of the CBE library interface) was a
dirty way of triggering the submit of a block-IO primitive for each completed
VBD primitive. This is now done internally during Library.Execute .

Changes to the CBE library interface
------------------------------------

:Obtain_Client_Data_2:

Removed.

Commit:
lib/cbe: get rid of address overlays

This removes all type casts via address overlays
(A : A_Type with Address => B'Address) and replaces them with calls to type
individual sub-programs that do a type-safe conversion from one type into the
other. Note, that this implies copying all the data on each conversion. The
conversion sub-programs all follow this interface pattern:

! procedure A_From_B (
!    A : out A_Type;
!    B :     B_Type);

The commit follows the most pedantic approach for all conversions. This might
not be desired for conversions between mere payload data types (e.g.
SHA256_4K_Data_From_CBE_Data). But the commit leaves such optimizations for the
future.

cnuke committed:
server/cbe_block: increase report buffer size

Depending on the tree topology the report will became huge. Accommodate
the current test case (where the VBD tree covers 32MiB).

cnuke committed:
server/cbe_block: show invalid block access

In case the tree (VBD and/or FT) is damaged to not fail silently but
report which node is broken.

cnuke committed:
lib/cbe: add Active function to Cache_Flusher

For now it is used to check if the flusher is currently doing
something.

cnuke committed:
lib/cbe: make sure to flush cache before securing

Before we update the superblock we have to make sure that all dirty
cache entries were written to disk. Therefor we check and instruct
the flusher to do its job first.

cnuke committed:
lib/cbe: flag new snapshot as valid

Was missed in 'cbe: get rid of invalid snapshot ID'.

cnuke committed:
lib/cbe: add To_String (Hash)

Add helper function to stringify the Hash_Type.

cnuke committed:
lib/cbe: add Next_Snap_Slot function

The whole approach to selecting the next snapshot slot probably has to
be changed WRT to verification. In there is no snapshot that can be
safely discarded user interventation is needed and therefor it is
possible for 'Cur_Snap = Next_Snap' to occur.

cnuke committed:
lib/cbe: unconditionally updaste snapshot PBA

There is no need to guard updating the snapshot information. The worst
case is that we update it needlessly.

cnuke committed:
server/cbe_block: mark execute progression

Reflect making progress to the event loop so that there is a chance to
process it once again.

cnuke committed:
lib/cbe: secure superblock on snapshot creation

Whenever a branch is changed we also have to update the superblock so
the list of active snapshots stays consistent. Not updating the super-
block was due to the fact the that the securing of the superblock is
actually a different operation than syncing, e.g. updating, the super-
block.

Commit:
cbe: gnatprove in mode "flow" succeeds

This especially cleans up initialization of variables prior to their use,
adjusts the modes of sub-program parameters ito their actual use, and avoids
unused function results. The CBE library is changed slightly.

Changes to the CBE library interface
------------------------------------

:Crypto_Cipher_Data_Required:

Library-object parameter has changed mode from "in out" to "in".

:Crypto_Plain_Data_Required:

Library-object parameter has changed mode from "in out" to "in".

cnuke committed:
server/cbe: print diagnostics only when option set

Add 'show_if_progress', i.e. print only when progress was made, option.

Commit:
cbe: clean up client interface

All sub-programs of the client interface of the CBE library now contain the
phrase "client" to make distinguishing the different CBE interfaces easier.
Also, the implementation of the sub-program for dropping completed client
requests was adjusted in the way that it now drops a specific request and not
any completed request as before.

Changes to the CBE library interface
------------------------------------

:Client_Request_Acceptable:

Renamed, former name was Request_Acceptable.

:Submit_Client_Request:

Renamed, former name was Submit_Request.

:Peek_Completed_Client_Request:

Renamed, former name was Peek_Completed_Request.

:Drop_Completed_Client_Request:

Renamed, former name was Drop_Completed_Request. Now receives the request that
shall be dropped as additional "in" parameter.

:Client_Data_Index:

Renamed, former name was Give_Data_Index.

cnuke committed:
server/cbe_block: add exit after init option

In case the 'cbe_block' is solely used to initialize the CBE make it
possible to exit afterwards. It then becomes possible for a staged
scenario to react upon that.

cnuke committed:
lib/cbe: remove cache flushing from SB sync

Creating a new snapshot will trigger the cache flush and other
operations, e.g. discarding snapshots, do not rely on a clean cache.

cnuke committed:
lib/cbe: add discard snapshot interface

Analog to the export interface this interface allows to remove a
snapshot from the list of active snapshots referenced by a super-
block.

Commit:
cbe: clean up tag type

The CBE tag type was a mod type that was used with several different meanings
inside the CBE:

1. As identifier of requests at the client side
2. As identifier of requests stored in the request pool
3. As coloring of a primitive (for instance associating module and operation)
4. As identifier of external events the CBE has to wait for
   (supply of client data, back-end block requests)

These meanings were mixed in values of the tag type leaving the prevention of
incorrectness or misinterpretation of such values merely to the discipline of
the programmer. This commit splits the former tag type up into four domain
specific types (corresponding to the use cases mentioned above) with tailored
value ranges.

* The new ``Request.Tag_Type`` is used only for the tag the CBE client assigned
  to a request. These values have no functional meaning for the CBE anymore as
  the client can chose them arbitrarily.

* The new ``Pool_Index_Type`` identifies a request at the request pool and has
  no invalid value.

* The new ``Pool_Index_Slot_Type`` is a pool index combined with a boolean
  whether the index is valid.

* In the new ``Primitive.Object_Type.Pool_Idx_Slot`` primitives carry the pool
  index of the request they correspond to when necessary. Note, that there are
  valid primitives that don't carry a pool index which is why the member must
  be a slot.

* The new ``Primitive.Tag_Type`` is only used for the coloring of primitives
  ("Write_Back", "Splitter", "Cache", ...). Note, that there is still an
  invalid value for this type as tags are stored outside of primitives. But
  the commits ensure that a valid primitive will now always contains a valid
  tag.

* The new ``Library.Event_Type`` identifies an external event that the CBE
  waits for to happen (for instance that the wrapper code supplies required
  client block-data). The value names of this type are very specific about the
  event that is expected and don't misuse the Tag_Type values anymore.

* The new ``Library.Object.Wait_For_Front/Back_End`` members replace the former
  Library.Object.Front/Back_End_Req_Prim and contain a member of Event_Type
  each. I hope this makes the inner workings of the CBE a little bit more
  clear.

Commit:
.ci/prove.mk: set mode via env variable PROVE_MODE

Default mode remains 'all'.

cnuke committed:
lib/cbe: remove Last_Snapshot_ID

Use generation in all places instead.

cnuke committed:
lib/cbe: only create snapshot when instructed

Only create the a new working snapshot (that is the current r/w
snapshot) when the superblock was secured as part of making a snapshot
persistent.

cnuke committed:
lib/cbe: use root PBA and hash for SN creation

This basic idea is to check last secured generation and root hash.
Since the working snapshot gen is last secured + 1 a sync will make
it persistent, just make sure that we really create the SN only when
necessary, e.g. the hash has changed.

(Rename the securing part to sync superblock b/c that's what it
really is and use securing for when we give the TA the hash of the
current SB is still pending.)

Commit:
lib/cbe: get rid of avoidable mod types

A lot of types in the CBE were still modular types eventhough it was not
necessary. Values of these types can wrap and the ranges are often much bigger
than required. Unused parts of such ranges are misused for mixing in
additional meanings. Furthermore, proving code is harder with modular types
and their flexibility. This commit replaces, as far as possible, all mod types
with types of a domain specific range. As a side precondition, values with
special meanings are replaced with additional variables of different type and
static constraints of the CBE are formalized.

cnuke committed:
lib/cbe: add SYNC request handling

A SYNC request will lead to storing the current working state of the
CBE in a new superblock. It, however, will not lead to securing the
generation.

(For now it is up the outside to differentiate the lastest superblock,
 interface (still) pending.)

cnuke committed:
lib/cbe: add queue frontend to Pool

The 'Index_Queue' is used to store the indices of the pool's requests
in FIFO order.

(Seperating the underlying pool item array from the index array is not
 strictly necessary, i.e. the item array could be modelled as queue
 directly.)

Commit:
lib/cbe: get rid of storage types

* Block_Size -> Block_Size_Bytes
* Type_I_Nodes_Per_Block
* make Type_I_Node_Type a normal type
* remove Type_1_Node_Info_Type
* Type_1_Node_Infos_Type -> Type_1_Node_Walk_Type
* index of Type_1_Node_Walk_Type is Tree_Level_Index_Type
* Infos -> Walk
* Get_*_Info -> Get_*_Node_Walk
* make Type_II_Node_Type a normal type
* Type_II_* -> Type_2_*
* Translation_Data_Type w/o Size assignm.
* CXX.Library.Initialize_Object receives SBs as block data
* make superblock and sub-types a normal type
* minor stuff, Hash_Index, Query_Data_Type
* get rid of Tree_Level_Invalid
* Height -> Max_Level
* get rid of Tree_Level_Type

Commit:
cbe_vfs.run: always rebuild Ada/SPARK objects

Remove CBE Ada objects in the build directory from previous builds.
This is done analogously to run/cbe.run to ensure that changes, for instance,
to ADS files are recognized on a rebuild (the toolchain doesn't ensure this
completely yet).

Commit:
lib/cbe_cxx: remove unused cbe-cxx.adb

cnuke committed:
cbe: use 64bit snapshot ids

Internally the generation was already used as external snapshot id.
However, the interface used a 32bit type that was directly used by
the C++ code. Now there is a proper type in place that contains the
64 bit value as well as an indication if the snapshot is in fact
valid.

cnuke committed:
run/cbe: increase parameters for server test

cnuke committed:
lib/cbe: store free-tree root hash in superblock

cnuke committed:
lib/cbe: guard against invalid requests

For now we raise an exception when we encounter an invalid request
where the count is 0.

cnuke committed:
run: add persistent storage run scripts

The 'cbe_init' run script will use the 'cbe_block' component to
initialize a image file with a clean CBE state. The default image size
is 512MiB and can be changed by setting the environment variable
'CBE_IMAGE_SIZE' (the value is given in MiB). Half of the image is used
for the initial VBD and the 8th part for the FT.

The 'cbe_test_compare' run script will override the first 4MiB of the
VBD twice while checking (i.e., the data that was written is read back
again) each block one by one.

cnuke committed:
lib/cbe: disable Execute debugging message

cnuke committed:
lib/cbe: remove usage of Superblocks types

For now there is only one in-memory copy of the Superblock in the CBE.

Commit:
lib/cbe: fix and rework Cache.Get_Cache_Item

* Rename Cache.Get_Cache_Item Unused_Or_Evictable_Item
* Move eviction code to new private Cache.Evictable_Item
* Do not evict items that are dirty
* Raise an exception if Cache.Unused_Or_Evictable_Item is not successful

jklmnn commented on src/lib/cbe/cbe-block_io.ads:
```diff
@@ -55,12 +56,16 @@ is
    --  \param Prim  reference to the Primitive
    --  \param Data  reference to a Block_data object
    --
+   procedure Submit_Primitive_Dont_Return_Index (
```
This could be an overloaded `Submit_Primitive`. That it doesn't return an index is clear from the signature. That it does the same as `Submit_Primitive` (which I would assume from the name anyway) can be part of the documentation.

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/bb18b5dad3d7a27c0862d3bd5cde232a249ff7e6.

jklmnn commented on src/lib/cbe/cbe-cache.ads:
```diff
@@ -48,11 +48,11 @@ is
-      Ts  :        Timestamp_Type)
-   return Cache_Index_Type;
+   procedure Data_Index (
+      Obj    : in out Object_Type;
```
The whitespace between the name and the `:` should be one space.

cnuke replied:
AFAICS we always align the `:` to the longset parameter, e.g.

```Ada
procedure Foo (
   First  : Type;
   Second : Type;
   Third  : Type)
```

I'm not sure if that is against the *Ada Style Guide* but in case it is not I would rather keep it that way.

jklmnn replied:
I somehow missed the `Result` paramter and thought there were only the three first ones.

jklmnn commented on src/lib/cbe/cbe-cache_flusher.adb:
```diff
@@ -102,6 +103,22 @@ is
+   --
+   --  Active
+   --
+   function Active (Obj : Object_Type)
+   return Boolean
+   is
+   begin
+      for Item_Id in Obj.Items'Range loop
+         if not Item.Invalid (Obj.Items (Item_Id)) then
+            return True;
+         end if;
+      end loop;
+
+      return False;
+   end Active;
```
This could be expressed as
```Ada
function Active (Obj : Object_Type) return Boolean is
   (for some I of Obj.Items => not Item.Invalid (I));
```

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/f687db0c858cb40e6fe6bcd4966b1063d2b46155.

jklmnn commented on src/lib/cbe/cbe-debug.ads:
```diff
@@ -0,0 +1,82 @@
+   with
+      SPARK_Mode => Off;
+
+   function To_String (Int : Uint64_Type)
```
Just as a note: The canonical name for the `To_String` function in Ada is already `Image` So it would make sense to remove this one and name all others `Image` (although this is just a hint, as this is only the debug package anyway).

cnuke replied:
Using the canonical name makes sense but I would post-pone changing that for now.

jklmnn commented on src/lib/cbe/cbe-library.adb:
```diff
@@ -84,82 +78,172 @@ is
    begin
-      Obj.Secure_Superblock := True;
-   end Start_Securing_Superblock;
+      Delcare_Cache_Flusher_Active :
```
This should probably have been `Declare_Cache_Flusher_Active`.

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/1bdc63891e10e5a3d5a3d3193cb10e4e910fda85.

jklmnn commented on src/lib/cbe/cbe-library.adb:
```diff
@@ -84,82 +78,172 @@ is
-   return Boolean
-   is (Obj.Seal_Generation);
+      Declare_Cache_Dirty :
+      declare
```
This declare is large enough to be a separate procedure.

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/52955e4b34f5c6bcd4cf7c31e060fa30fcc2d737.

jklmnn commented on src/lib/cbe/cbe-library.adb:
```diff
@@ -192,115 +275,175 @@ is
+      if Primitive.Valid (Obj.Sync_Primitive) and then
+         not Obj.Secure_Superblock
+      then
+         Delcare_Sync_Cache_Flusher_Active :
```
Again, typo, but this declare could also be a separate procedure.

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/2ac150090eb682102796877b0932e115fc227357.

jklmnn commented on src/lib/cbe/cbe-library.adb:
```diff
@@ -192,115 +275,175 @@ is
+      Write_Back.Peek_Completed_Root_Hash (WB, Prim, Snap.Hash);
    end Update_Snapshot_Hash;
 
    procedure Execute (
```
In general, `Execute` is already splitted into multiple logical blocks. Each of these blocks could be implemented by a separate procedure instead of a declare block.

cnuke replied:
You are right, it's on the todo list already - I'll open an issue for that (see #38).

jklmnn commented on src/lib/cbe/cbe-library.ads:
```diff
@@ -228,12 +219,12 @@ is
-      Data :        Block_Data_Type)
-   return Boolean;
+   procedure Supply_Client_Data (
+      Obj      : in out Object_Type;
```
Whitespace between name and `:`.

cnuke replied:
See first comment.

jklmnn commented on src/lib/cbe/cbe-pool.adb:
```diff
@@ -121,6 +129,10 @@ is
+      if Index_Queue.Full (Obj.Indices) then
+         return False;
+      end if;
+
```
`Request_Acceptable` could be expressed as
```Ada
function Request_Acceptable (Obj : Object_Type) return Boolean is
   (not Index_Queue.Full (Obj.Indices) and then (for some I of Obj.Items => Item.Invalid (I));
```

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/bb40fe3075a0e63d2b19a553d2a97a3c981038e6.

jklmnn commented on src/lib/cbe/cbe-virtual_block_device.adb:
```diff
@@ -247,26 +222,23 @@ is
-                  end Declare_Data;
+                     Cache.Data_Index (Cach, PBA, Timestamp, Data_Index);
+                     Declare_Data :
+                     declare
```
Is a separate declare really necessary here?

cnuke replied:
We could make `Data` non-constant and declare it in the `Data_Index` block but I'm not sure if that's worthwhile. I *think* one of the reasons for making it constant could have been to force the compiler to not create a copy but that may have been futile after all. In this case it probably would be better to change the `Translation` module interface and pass the whole cache data as well as the index in - I've to think about that.

jklmnn commented on src/lib/cbe/cbe.adb:
```diff
@@ -8,79 +8,506 @@
+      Key_Off : Block_Data_Index_Type := Data_Off;
+   begin
+      Declare_Value_Off : declare
+         Value_Off : Block_Data_Index_Type;
```
This declaration could be moved to the declare section of the procedure and the declare block could be removed.

cnuke replied:
Done.

jklmnn replied:
Fixed by https://github.com/cbe-project/cbe/pull/37/commits/3dba3166a3e5cfc0e195fb084b95eef0c6e19cd9.

cnuke committed:
lib/cbe: add Dump_Cache_State procedure

Commit:
lib/cbe: style errors in cbe-cxx.ads

Commit:
lib/cbe: hardcode layout of metadata blocks

The free-number member in the superblock type was written to a bad offset
inside the block when converting superblocks to data blocks. This was because
we used 'Size attributes to calculate the offset. One of the types we used with
'Size was changed over time from a storage to a normal range type and therefore
changed in size. The layout of metadata blocks should be hardcoded and not
depend on implementation details of components that use it.

cnuke committed:
server/cbe_block: properly report type 2 nodes

cnuke committed:
lib/cbe: change when generation is set

cnuke committed:
lib/cbe: use cache in free-tree

cnuke committed:
cbe: print superblock

cnuke committed:
lib/cbe: use next SB slot + change root node check

cnuke committed:
lib/cbe: fix active snapshot list binding

cnuke committed:
server/cbe_block: catch broken block requests

Commit:
server/cbe: avoid undefined reference to memcmp

cnuke committed:
lib/cbe: fix FT level range

The range is [1, height+1].

cnuke committed:
lib/cbe: adapt to Tree_Level_Index_Type

jklmnn commented:
@cnuke what is the current state of the changes that were addressed in this pull request?

cnuke commented:
@jklmnn ah sorry, I deleted my staging branch w/o think about what might happen to the PR. Your remarks were noted but so far not addressed.
